### PR TITLE
GSOC UPGRADING TO DSTU3 VERSION

### PR DESCRIPTION
**Pull request by kwateng**

### CODE REVIEW

**Comment by dkayiwa** [comment]
Do you mind squashing the commits into one as advised at? https://wiki.openmrs.org/display/docs/Pull+Request+Tips

**Comment by kwateng** [comment]
Thank you very much Dkayiwa. This is still pending. Lots of API Changes in STU3. Will do when I finish it.

**Comment by dkayiwa** [comment]
@kwateng as you said that this is work in progress, do you ping your mentor to review it?

**Comment by kwateng** [comment]
@dkayiwa I had detailed conversation with @harsha89. Now all compilation errors sorted. Mentor asked to do comprehensive testing as this is a major change for the module. I'm doing it. Informed @harsha89 to review this and you can review this too.

**Comment by harsha89** [comment]
@dkayiwa I will review. We need a lot of testing due to major version upgrade. If you have a time please review this. I will do it today or tomorrow.

File: api/src/main/java/org/openmrs/module/fhir/api/impl/EncounterServiceImpl.java [hunk 1]
```diff
@@ -134,7 +139,12 @@ public Encounter getEncounter(String id) {
 		org.openmrs.Encounter omrsEncounter = Context.getEncounterService().getEncounterByUuid(id);
 		List<Composition> encounterList = new ArrayList<Composition>();
 		if (omrsEncounter != null) {
-			encounterList.add(FHIREncounterUtil.generateComposition(omrsEncounter));
+			try {
+				encounterList.add(FHIREncounterUtil.generateComposition(omrsEncounter));
+			} catch (FHIRException e) {
+				String msg = "Error while searching the encounter composition";
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/impl/EncounterServiceImpl.java]: move this to a constant

File: api/src/main/java/org/openmrs/module/fhir/api/impl/EncounterServiceImpl.java [hunk 2]
```diff
@@ -147,7 +157,12 @@ public Encounter getEncounter(String id) {
 		List<org.openmrs.Encounter> omrsEncounters = Context.getEncounterService().getEncountersByPatient(patient);
 		List<Composition> fhirEncounters = new ArrayList<Composition>();
 		for (org.openmrs.Encounter enc : omrsEncounters) {
-			fhirEncounters.add(FHIREncounterUtil.generateComposition(enc));
+			try {
+				fhirEncounters.add(FHIREncounterUtil.generateComposition(enc));
+			} catch (FHIRException e) {
+				String msg = "Error while searching the encounter composition by patient id";
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/impl/EncounterServiceImpl.java]: samehere

File: api/src/main/java/org/openmrs/module/fhir/api/impl/EncounterServiceImpl.java [hunk 3]
```diff
@@ -159,7 +174,12 @@ public Encounter getEncounter(String id) {
 		org.openmrs.Encounter omrsEncounter = Context.getEncounterService().getEncounterByUuid(encounterId);
 		List<Composition> encounterList = new ArrayList<Composition>();
 		if (omrsEncounter != null) {
-			encounterList.add(FHIREncounterUtil.generateComposition(omrsEncounter));
+			try {
+				encounterList.add(FHIREncounterUtil.generateComposition(omrsEncounter));
+			} catch (FHIRException e) {
+				String msg = "Error while searching the encounter composition by encounter id";
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/impl/EncounterServiceImpl.java]: same here

File: api/src/main/java/org/openmrs/module/fhir/api/impl/EncounterServiceImpl.java [hunk 4]
```diff
@@ -217,7 +237,10 @@ public void deleteEncounter(String id) {
 		if (encounter == null) {
 			Visit visit = Context.getVisitService().getVisitByUuid(id);
 			if (visit == null) {
-				throw new ResourceNotFoundException(Encounter.class, new IdDt("Encounter", id));
+				//Jira related https://issues.openmrs.org/browse/FM-194
+				IIdType idType = new IdDt();
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/impl/EncounterServiceImpl.java]: is this should be IdDt I saw we are using IDType now

File: api/src/main/java/org/openmrs/module/fhir/api/impl/PatientServiceImpl.java [hunk 5]
```diff
@@ -282,7 +282,8 @@ public void deletePatient(String id) {
 		org.openmrs.Patient patient = Context.getPatientService().getPatientByUuid(id);
 		// patient not found. return with 404
 		if (patient == null) {
-			throw new ResourceNotFoundException(Patient.class, new IdDt("Patient", id));
+			//Jira related https://issues.openmrs.org/browse/FM-194
+			throw new ResourceNotFoundException(new IdDt("Patient", id));
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/impl/PatientServiceImpl.java]: Spuild we ise IDType here instead?

File: omod/src/main/java/org/openmrs/module/fhir/providers/RestfulLocationResourceProvider.java [hunk 6]
```diff
@@ -117,11 +121,14 @@ public void deleteLocation(@IdParam IdDt theId) {
...
-	public MethodOutcome updateLocation(@ResourceParam Location location, @IdParam IdDt theId) {
+	public MethodOutcome updateLocation(@ResourceParam Location location, @IdParam IdType theId) {
 		MethodOutcome retVal = new MethodOutcome();
 		OperationOutcome outcome = new OperationOutcome();
-		location = locationResource.updateLocation(theId.getIdPart(), location);
-		outcome.addIssue().setDetails("Location successfully updated");
+		locationResource.updateLocation(theId.getIdPart(), location);
+		CodeableConcept concept = new CodeableConcept();
+		Coding coding = concept.addCoding();
+		coding.setDisplay("Location successfully updated");
```
Comment (harsha89) [inline omod/src/main/java/org/openmrs/module/fhir/providers/RestfulLocationResourceProvider.java]: move to a constant. we shpuld include the Id as well?

File: omod/src/main/java/org/openmrs/module/fhir/providers/RestfulLocationResourceProvider.java [hunk 7]
```diff
@@ -177,9 +189,12 @@ public MethodOutcome updateLocationByName(@ResourceParam Location theLocation, @
 	public MethodOutcome createFHIRLocation(@ResourceParam Location location) {
 		location = locationResource.createLocation(location);
 		MethodOutcome retVal = new MethodOutcome();
-		retVal.setId(new IdDt(FHIRConstants.LOCATION, location.getId().getIdPart()));
+		retVal.setId(new IdDt(FHIRConstants.LOCATION, location.getId()));
 		OperationOutcome outcome = new OperationOutcome();
-		outcome.addIssue().setDetails("Location is successfully created");
+		CodeableConcept concept = new CodeableConcept();
+		Coding coding = concept.addCoding();
+		coding.setDisplay("Location is successfully created");
```
Comment (harsha89) [inline omod/src/main/java/org/openmrs/module/fhir/providers/RestfulLocationResourceProvider.java]: May be we need to mention whoch resource created by gibing the id as well

**Comment by kwateng** [comment]
@harsha89 I added fixes when I do the testing. Regarding moving to constants. Those sentences used by one time. And I add Id to them. So I don't think we need to move to constants.

**Comment by harsha89** [comment]
Alright, can you follow the document which shared by @dkayiwa and squashes the commits?

**Comment by kwateng** [comment]
@harsha89 completed

**Comment by harsha89** [comment]
@kwateng I have merged the PR. Please create a ticket for this.

**Comment by kwateng** [comment]
@harsha89 added jira https://issues.openmrs.org/browse/FM-199

**Pull request merged**